controllers/application/network_policy.go

```diff
@@ -2,61 +2,53 @@ package applicationcontroller
 	"github.com/kartverket/skiperator/pkg/util"
-	corev1 "k8s.io/api/core/v1"
 	networkingv1 "k8s.io/api/networking/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/intstr"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-const (
-	GrafanaAgentName      = "grafana-agent"
-	GrafanaAgentNamespace = GrafanaAgentName
-)
-
-// This is a bit hacky, but seems like best solution
-func (r *ApplicationReconciler) NetworkPoliciesFromService(ctx context.Context, obj client.Object) []reconcile.Request {
-	svc := obj.(*corev1.Service)
+func (r *ApplicationReconciler) reconcileNetworkPolicy(ctx context.Context, application *skiperatorv1alpha1.Application) (reconcile.Result, error) {
+	controllerName := "NetworkPolicy"
+	r.SetControllerProgressing(ctx, application, controllerName)
 
-	applications := &skiperatorv1alpha1.ApplicationList{}
-	err := r.GetClient().List(ctx, applications)
+	egressServices, err := r.GetEgressServices(ctx, application, application.Spec.AccessPolicy)
 	if err != nil {
-		return nil
+		r.SetControllerFinishedOutcome(ctx, application, controllerName, err)
+		return reconcile.Result{}, err
 	}
 
-	requests := make([]reconcile.Request, 0, len(applications.Items))
-	for _, application := range applications.Items {
-		if application.Spec.AccessPolicy == nil {
-			continue
-		}
-		for _, rule := range application.Spec.AccessPolicy.Outbound.Rules {
-			if rule.Namespace == svc.Namespace && rule.Application == svc.Name {
-				requests = append(requests, reconcile.Request{
-					NamespacedName: types.NamespacedName{
-						Namespace: application.Namespace,
-						Name:      application.Name,
-					},
-				})
-				break
-			}
-		}
+	networkPolicy := networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: application.Namespace,
+			Name:      application.Name,
+		},
 	}
-	return requests
-}
 
-func (r *ApplicationReconciler) reconcileNetworkPolicy(ctx context.Context, application *skiperatorv1alpha1.Application) (reconcile.Result, error) {
-	controllerName := "NetworkPolicy"
-	r.SetControllerProgressing(ctx, application, controllerName)
+	netpolSpec := networking.CreateNetPolSpec(
+		networking.NetPolOpts{
+			AccessPolicy:     application.Spec.AccessPolicy,
+			Ingresses:        &application.Spec.Ingresses,
+			Port:             &application.Spec.Port,
+			Namespace:        application.Namespace,
+			Name:             application.Name,
+			RelatedServices:  &egressServices,
+			PrometheusConfig: application.Spec.Prometheus,
+		},
+	)
 
-	networkPolicy := networkingv1.NetworkPolicy{ObjectMeta: metav1.ObjectMeta{Namespace: application.Namespace, Name: application.Name}}
-	_, err := ctrlutil.CreateOrPatch(ctx, r.GetClient(), &networkPolicy, func() error {
+	if netpolSpec == nil {
```

Comment:
Have you taken into account the case where a user goes from having a networkpolicy to not having one? In other words, should there be a delete of orphaned netpols here?

Reply:
Not sure I follow?

In the case that someone has a NetPol, but changes the Application in a way which makes the NetPol Spec empty, the original NetPol will be deleted. If the NetPol spec is not empty, it will be created by the `CreateOrPatch` function below

Reply:
The first case you mentioned is what I'm referring to. Where the user used to have a netpol, but has changed the config so that it no longer should have one.

I see it now, it's nested as an argument to the function call below, so this seems fine. It might cause a few unecessary deletes though, GETs are generally cheaper as they can be cached. But that probably won't matter in the big picture.

```
r.GetClient().Delete(ctx, &networkPolicy)
```

Reply:
If you like I can put it in two steps, so it's easier to read what's going on :)